rust/kernel/sync/condvar.rs

```diff
@@ -44,15 +58,27 @@ unsafe impl Sync for CondVar {}
 
 impl CondVar {
     /// Constructs a new conditional variable.
-    ///
-    /// # Safety
-    ///
-    /// The caller must call `CondVar::init` before using the conditional variable.
-    pub const unsafe fn new() -> Self {
-        Self {
-            wait_list: Opaque::uninit(),
-            _pin: PhantomPinned,
+    #[allow(clippy::new_ret_no_self)]
+    pub const fn new(name: &'static CStr, key: &'static LockClassKey) -> impl PinInit<Self> {
+        fn init_wait_list(
```

Comment:
What's the rationale behind having a new function instead of just having
```rust
wait_list: unsafe { init::pin_init_from_closure(|| { ... }) }
```
down there?

Reply:
here would be the closure version:
```rust
        pin_init!(Self {
            wait_list: unsafe {
                init::pin_init_from_closure::<_, core::convert::Infallible>(move |slot| {
                    bindings::__init_waitqueue_head(
                        Opaque::raw_get(slot),
                        name.as_char_ptr(),
                        key.get(),
                    );
                    Ok(())
                })
            },
            _pin: PhantomPinned,
        })
```
(yes we need to specify the error type 🙄)

Not sure if that is more readable... I could probably create a more general function that would handle this, so along the lines of:
```rust
unsafe fn ffi_init1<T, P1>(fun: unsafe fn(*mut T, P1), param1: P1) -> impl PinInit<Opaque<T>>;
unsafe fn ffi_init2<T, P1, P2>(fun: unsafe fn(*mut T, P1), param1: P1, param2: P2) -> impl PinInit<Opaque<T>>;
unsafe fn ffi_init3<T, P1, P2, P2>(fun: unsafe fn(*mut T, P1), param1: P1, param2: P2, param3: P3) -> impl PinInit<Opaque<T>>;
```